apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data/PathUtilities.cs

```diff
@@ -0,0 +1,84 @@
+﻿// Copyright 2017, Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using System;
+using System.Linq;
+using System.Security.Cryptography;
+
+namespace Google.Cloud.Firestore.Data
+{
+    internal static class PathUtilities
+    {
+        private static readonly string s_generationDomain = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
+        private static readonly RandomNumberGenerator s_rng = RandomNumberGenerator.Create();
+        private static readonly byte[] s_rngBuffer = new byte[30];
+
+        private static readonly char[] s_slashSplit = new[] { '/' };
+        /// <summary>
+        /// Splits a path by slashes, and validates that no element is empty.
+        /// </summary>
+        /// <param name="path">Path to split. Must not be null.</param>
+        /// <returns>An array of path elements, all of which are non-empty.</returns>
+        internal static string[] SplitPath(string path)
+        {
+            GaxPreconditions.CheckNotNull(path, nameof(path));
+            string[] elements = path.Split(s_slashSplit);
+            if (elements.Contains(""))
+            {
+                throw new ArgumentException("Path cannot contain empty elements", nameof(path));
+            }
+            return elements;
+        }
+
+        internal static string ValidateId(string id, string paramName = "id")
+        {
+            GaxPreconditions.CheckNotNullOrEmpty(id, paramName);
+            GaxPreconditions.CheckArgument(!id.Contains('/'), paramName, "ID cannot contain a '/' character.");
+            return id;
+        }
+
+        /// <summary>
+        /// Generates a random ID consisting of exactly 20 characters, each in the range a-z, A-Z, 0-9.
+        /// </summary>
+        internal static string GenerateId()
+        {
+            char[] chars = new char[20];
+            lock (s_rngBuffer)
```

Comment:
Would be simpler to simply create a new local rngBuffer here, rather than have global one and lock.
I can't quite believe performance matters that much? Mind you, is locking faster than a single small allocation anyway?

Reply:
Well I'd want to lock anyway, to avoid using RandomNumberGenerator from multiple threads concurrently. At that point, we might as well reuse a single buffer.